aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h

```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+    output_dims[1] = compute_output_dimension(input_dims[1],  // height
+                                              pad[0] + pad[2],
+                                              /*adjustment=*/0,
+                                              kernel_dims[1],
+                                              dilation[1],
+                                              subsampling_dims[1],
+                                              /*transpose=*/false);
+    return output_dims;
+  }
+};
+
+struct deconv_param_t {
+  const std::array<uint32_t, 2> kernel_dims; // kernel width, kernel height
+  const std::array<uint32_t, 2> stride_dims; // stride width, height
+  const std::array<uint32_t, 2> dilation_dims; // dilation width, height
+  const std::array<uint32_t, 4> padding_dims; // input padding top, left, bottom, right
+  const std::array<uint32_t, 2> adjustment_dims; // output adjustment
+
+  const uint32_t groups;
+  const size_t input_channels;
+  const size_t output_channels;
+  const uint8_t kernel_zero_point;
+  const float kernel_scale;
+  const uint8_t output_min;
+  const uint8_t output_max;
+
+  // The following are derived parameters
+  enum pytorch_qnnp_ukernel_type ukernel_type; // kernel type based on input params
+  size_t group_input_channels;
+  size_t group_output_channels;
+
+  /**
+   * @brief Constructor for initializing the deconvolution parameters.
+   */
+  deconv_param_t(
+      const std::array<uint32_t, 2> kernel,
+      const std::array<uint32_t, 2> stride,
+      const std::array<uint32_t, 2> dil,
+      const std::array<uint32_t, 4> pd,
+      const std::array<uint32_t, 2> adj,
+
+      const uint32_t grp,
+      const size_t in_ch,
+      const size_t out_ch,
+      const uint8_t kernel_zp,
+      const float kernel_s,
+      const uint8_t out_min,
+      const uint8_t out_max)
+      : kernel_dims(kernel),
+        stride_dims(stride),
+        dilation_dims(dil),
+        padding_dims(pd),
+        adjustment_dims(adj),
+        groups(grp),
+        input_channels(in_ch),
+        output_channels(out_ch),
+        kernel_zero_point(kernel_zp),
+        kernel_scale(kernel_s),
+        output_min(out_min),
+        output_max(out_max) {
+
+    const uint32_t kernel_width = kernel_dims[0];
+    const uint32_t kernel_height = kernel_dims[1];
+    const uint32_t dilation_width = dilation_dims[0];
+    const uint32_t dilation_height = dilation_dims[1];
+    const uint32_t stride_width = stride_dims[0];
+    const uint32_t stride_height = stride_dims[1];
+    const uint32_t adjustment_width = adjustment_dims[0];
+    const uint32_t adjustment_height = adjustment_dims[1];
+
+    const uint32_t input_padding_top = padding_dims[0];
+    const uint32_t input_padding_left = padding_dims[1];
+    const uint32_t input_padding_bottom = padding_dims[2];
+    const uint32_t input_padding_right = padding_dims[3];
+
+    group_input_channels = input_channels / groups;
```

Comment:
We should add a check here for groups > 0 and input_channels % groups == 0? This is common to both conv and deconv. If you can merge them that would be good to, although I know we decided to defer this to a later PR. But this might just reduce the places that needs such fixing.